crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -553,6 +553,93 @@ impl<'w, 's, T: Component> Fetch<'w, 's> for WriteFetch<T> {
     }
 }
 
+pub struct Has<T: Component>(PhantomData<T>);
+
+impl<T: Component> WorldQuery for Has<T> {
+    type Fetch = HasFetch<T>;
+    type State = HasState<T>;
+}
+
+pub struct HasFetch<T: Component> {
+    matches: bool,
+    _marker: PhantomData<T>,
+}
+
+/// SAFETY: HasFetch is read only because it fetches no data
+unsafe impl<T: Component> ReadOnlyFetch for HasFetch<T> {}
+
+#[derive(Clone)]
+pub struct HasState<T: Component> {
+    component_id: ComponentId,
+    _marker: PhantomData<T>,
+}
+
+unsafe impl<T: Component> FetchState for HasState<T> {
+    fn init(world: &mut World) -> Self {
+        Self {
+            component_id: world.init_component::<T>(),
+            _marker: PhantomData,
+        }
+    }
+
+    fn update_component_access(&self, _: &mut FilteredAccess<ComponentId>) {}
+
+    fn update_archetype_component_access(
+        &self,
+        _: &Archetype,
+        _: &mut Access<ArchetypeComponentId>,
+    ) {
+    }
+
+    fn matches_archetype(&self, archetype: &Archetype) -> bool {
+        archetype.contains(self.component_id)
```

Comment:
Isn't the intention that all entities match, but a bool is returned if it has a specific component?

Reply:
Yeah, the option fetch unconditionally returns true here